models/exposure.py

```diff
@@ -1,7 +1,238 @@
+                break
+
+    def __repr__(self):
+        return f"Exposure(id: {self.id}, exp: {self.exp_time}s, filt: {self.filter}, from: {self.instrument}/{self.telescope})"
+
+    def __str__(self):
+        return self.__repr__()
+
+    def save(self):
+        pass  # TODO: implement this! do we need this?
+
+    def load(self, ccd_ids=None):
+        if ccd_ids is None:
+            ccd_ids = list(range(self.num_ccds))
+
+        if not isinstance(ccd_ids, list):
+            ccd_ids = [ccd_ids]
+
+        if not all([isinstance(ccd_id, int) for ccd_id in ccd_ids]):
+            raise ValueError("ccd_ids must be a list of integers. ")
+
+        if self.filename is not None:
+            for i in ccd_ids:
+                self.data[i]  # use the CCD_Data __getitem__ method to load the data
+        else:
+            raise ValueError("Cannot load data from database without a filename! ")
+
+    @property
+    def data(self):
+        if self._data is None:
+            inst = self.get_instrument_object()
+            if inst is None:
+                if self.instrument is None:
+                    raise ValueError("Cannot have an instrument with no name! ")
+                else:
+                    raise ValueError(f"Could not find instrument with name: {self.instrument}! ")
+            self._data = CCD_Data(self.filename, inst)
+        return self._data
+
+    @data.setter
+    def data(self, value):
+        if not isinstance(value, CCD_Data):
+            raise ValueError(f"data must be a CCD_Data object. Got {type(value)} instead. ")
+        self._data = value
+
+    def get_instrument_object(self):
+        if self.instrument_object is not None:
+            if self.instrument is None:
+                self.instrument_object = None
+
+            else:
+                self.instrument_object = getattr(models.instruments, self.instrument)()
+
+        return self.instrument_object
+
+    def calculate_coordinates(self):
+        if self.ra is None or self.dec is None:
+            raise ValueError("Exposure must have RA and Dec set before calculating coordinates! ")
+
+        coords = SkyCoord(self.ra, self.dec, unit="deg", frame="icrs")
+        self.gallat = coords.galactic.b.deg
+        self.gallon = coords.galactic.l.deg
+        self.ecllat = coords.barycentrictrueecliptic.lat.deg
+        self.ecllon = coords.barycentrictrueecliptic.lon.deg
+
+    def read_header(self):
+        inst = self.get_instrument_object()
+
+        # read header info, put it in the header JOSNB column
+        if inst is not None:
+            self.header = inst.read_header(self.filename)
+
+    def parse_header_keys(self):
+        """
+        Parse the relevant columns: mjd, project, target,
+        num_ccds, width, height, exp_time, filter, telescope
+        from self.header and into the column attributes.
+        """
+
+        for k, v in self.header:
+            norm_k = normalize_header_key(k)
```

Comment:
This can't be in a general pipeline.utils -- it needs to be something that's in the instrument class.

If we try to write one set of code that works for all instruments, we will hate ourselves later.  Much cleaner just to have the defaults in an instrument superclass, and then have the right answer for a given instrument (if it's not the default) in the instrument subclass.

Look at : https://gitlab.com/robknop/curveball/-/blob/master/bin/exposuresource.py  Look at the properties starting at `ra_degrees_keyword`.  (I override those properties in subclasses of the ExposureSource class as necessary.)

Reply:
This just makes it easier to match keywords. I don't think there's anything instrument specific about it. Right now it just removes spaces and underscores and puts it in upper case. Think of it as a more generic `.upper()` function, to be used for matching the keys. 
The list of keywords that get matched could be a different story. I guess we should allow each instrument object to define column names for those fields.